libc/stdio_scanf.c

```diff
@@ -0,0 +1,4371 @@
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+
+	const char *buff = "loremIPSUM IPSUMdolor dolorSitAmet";
+	const char *buff2 = "123loremIPSUM IPSUMdolor123 dolor123SitAmet";
+	char *res1, *res2, *res3;
+	char *format = "%m[A-z] %m[A-Z] %m[a-z]";
+
+	fprintf(filep, "%s", buff);
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(3, fscanf(filep, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM", res1);
+	TEST_ASSERT_EQUAL_STRING("IPSUM", res2);
+	TEST_ASSERT_EQUAL_STRING("dolor", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(3, test_vfscanfWrapper(filep, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM", res1);
+	TEST_ASSERT_EQUAL_STRING("IPSUM", res2);
+	TEST_ASSERT_EQUAL_STRING("dolor", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	TEST_ASSERT_EQUAL_INT(3, test_vsscanfWrapper(buff, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM", res1);
+	TEST_ASSERT_EQUAL_STRING("IPSUM", res2);
+	TEST_ASSERT_EQUAL_STRING("dolor", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	TEST_ASSERT_EQUAL_INT(3, sscanf(buff, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM", res1);
+	TEST_ASSERT_EQUAL_STRING("IPSUM", res2);
+	TEST_ASSERT_EQUAL_STRING("dolor", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	fclose(filep);
+	filep = fopen(TESTFILE_PATH, "w+");
+	format = "%m[1-9] %m[^1-9] %m[1-9]";
+
+	fprintf(filep, "%s", buff2);
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(3, fscanf(filep, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("123", res1);
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM IPSUMdolor", res2);
+	TEST_ASSERT_EQUAL_STRING("123", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	rewind(filep);
+
+	TEST_ASSERT_EQUAL_INT(3, test_vfscanfWrapper(filep, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("123", res1);
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM IPSUMdolor", res2);
+	TEST_ASSERT_EQUAL_STRING("123", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	TEST_ASSERT_EQUAL_INT(3, test_vsscanfWrapper(buff2, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("123", res1);
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM IPSUMdolor", res2);
+	TEST_ASSERT_EQUAL_STRING("123", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+
+	TEST_ASSERT_EQUAL_INT(3, sscanf(buff2, format, &res1, &res2, &res3));
+	TEST_ASSERT_EQUAL_STRING("123", res1);
+	TEST_ASSERT_EQUAL_STRING("loremIPSUM IPSUMdolor", res2);
+	TEST_ASSERT_EQUAL_STRING("123", res3);
+	free(res1);
+	free(res2);
+	free(res3);
+}
+
+
+TEST(stdio_scanf_rest, m_c)
+{
+/* Disabled because of issue: https://github.com/phoenix-rtos/phoenix-rtos-project/issues/667 */
+#ifdef __phoenix__
+	TEST_IGNORE();
+#endif
+
+/* Address sanitizer used on Ubuntu 22.04 fails in such case, because of the following issue:
+ * https://github.com/llvm/llvm-project/issues/61768
+ */
+#ifndef __phoenix__
```

Comment:
IMHO would be better to disable this test only if AddressSanitizer is enabled.
Something like: 
```c
#if __has_feature(address_sanitizer) || defined(__SANITIZE_ADDRESS__)
```

(taken from https://github.com/llvm/llvm-project/blob/1a2eaebc09c6a200f93b8beb37130c8b8aab3934/compiler-rt/include/sanitizer/asan_interface.h#L51)

Reply:
There are some problems with __has_feature but is easy to go over it using only ifdef no_SANITIZE_ADDRESS and checking if his value is 1, which means is active

Reply:
isn't `defined(__SANITIZE_ADDRESS__)` enough? That's how the ASAN internally checks if it's enabled.

if not, please make it a single line
`#if defined(__SANITIZE_ADDRESS__) && (__SANITIZE_ADDRESS__ == 1)`

Reply:
You were right just `#if defined(__SANITIZE_ADDRESS__)` was enough.